### PR TITLE
[NP-3576] - GenericBusinessValidator ViewReference Update

### PR DESCRIPTION
**Pull request by nanopay-arthur**
Refs: 
https://nanopay.atlassian.net/browse/NP-3590
https://nanopay.atlassian.net/browse/NP-3576
https://nanopay.atlassian.net/browse/NP-3578

**Approvals Menu for Treviso Fraud-Ops Group**
Screen Shot 2021-03-21 at 10 07 04 PM
Screen Shot 2021-03-21 at 10 08 35 PM

**Generic Business Validator View Reference View**
Screen Shot 2021-03-21 at 10 07 14 PM
Screen Shot 2021-03-21 at 10 08 56 PM

**Saving with Fully Valid Data**
Screen Shot 2021-03-21 at 10 07 33 PM
Screen Shot 2021-03-21 at 10 09 10 PM

**Saving with Invalid Data**
Screen Shot 2021-03-21 at 10 07 53 PM
Screen Shot 2021-03-21 at 10 09 24 PM

### CODE REVIEW

File: src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.crunch',
+  name: 'AgentCapabilityJunctionDAOSummaryView',
```
Comment (jlhughes) [inline src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js:9]: This view is specific to Approvals and as such should be in the approval package. 
Also there is reference to a nanopay classification.

File: tools/classes.js [hunk 2]
```diff
@@ -797,6 +797,7 @@ var classes = [
   'foam.nanos.crunch.CapabilityCategoryCapabilityJunction',
   'foam.nanos.crunch.CapabilityJunctionStatus',
   'foam.nanos.crunch.UserCapabilityJunction',
+  'foam.u2.wizard.data.BooleanHolder',
```
Comment (jlhughes) [inline tools/classes.js]: Please remove from this particular PR.  The holders will change package.

File: src/foam/nanos/crunch/ruler/ConfigureUCJExpiryOnGranted.js [hunk 3]
```diff
@@ -79,7 +79,6 @@ foam.CLASS({
             logger.debug(this.getClass().getSimpleName(), "ucj.setExpiry()", ucj);
             if ( junctionExpiry != null && ( data instanceof RenewableData ) ) {
               ((RenewableData) data).setRenewable(false);
-              ((RenewableData) data).setReviewed(false);
```
Comment (Nauna) [inline src/foam/nanos/crunch/ruler/ConfigureUCJExpiryOnGranted.js:82]: Just commenting that I agree with this. The configureUcjExpiryOnGranted - should not be specific to one capability.

File: src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js [hunk 4]
```diff
@@ -18,15 +18,26 @@ foam.CLASS({
...
     function save(wizardlet) {
       if ( ! wizardlet.isAvailable ) return Promise.resolve();
+      var wData = wizardlet.data ? wizardlet.data.clone() : null;
       wizardlet.loading = true;
-      return this.crunchService.updateJunction( null,
-        wizardlet.capability.id,
-        wizardlet.data ? wizardlet.data.clone() : null,
-        null
-      ).then((ucj) => {
+      let p = this.subject ? this.crunchService.updateJunctionFor(
```
Comment (Nauna) [inline src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js:38]: when would the subject exist and no exist?
I'd recommend a comment - not specifically mine - this is just a suggestion
```suggestion
      // Generic model used in backoffice & user application. Subject override should only be hit if backoffice
      let p = this.subject ? this.crunchService.updateJunctionFor(
```

File: src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js [hunk 5]
```diff
@@ -42,9 +53,12 @@ foam.CLASS({
     },
     function load(wizardlet) {
       wizardlet.loading = true;
-      return this.crunchService.getJunction(
+      let p = this.subject ? this.crunchService.getJunctionFor(
```
Comment (Nauna) [inline src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js:60]: I guess to be consistent, the same comment - if you agreed to add - should be here too.

File: src/foam/u2/crunch/wizardflow/LoadCapabilitiesAgent.js [hunk 6]
```diff
@@ -41,19 +41,28 @@ foam.CLASS({
...
+          .then(capabilities => { this.capabilities = capabilities });
+      }
       return this.crunchService.getCapabilityPath(null, this.rootCapability.id, false)
         .then(capabilities => { this.capabilities = capabilities });
     },
     function getWAO() {
       switch ( this.waoSetting ) {
         case this.WAOSetting.UCJ:
-          return this.UserCapabilityJunctionWAO.create({}, this.__context__);
+          return this.UserCapabilityJunctionWAO.create({ subject: this.subject ? this.subject : null }, this.__context__);
```
Comment (Nauna) [inline src/foam/u2/crunch/wizardflow/LoadCapabilitiesAgent.js]: is this really necessary. If subject is falsy shouldn't UserCapabilityJunctionWAO know that - does it really have to pass back a null?

**Review by Nauna — APPROVED** [review]
haven't test but logically makes sense

File: tools/classes.js (hunk 2 above)
Comment (nanopay-arthur) [reply tools/classes.js]: Addressed.

File: src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js (hunk 4 above)
Comment (nanopay-arthur) [reply src/foam/nanos/crunch/ui/UserCapabilityJunctionWAO.js:38]: Added documentation to subject property for clarification.

File: src/foam/u2/crunch/wizardflow/LoadCapabilitiesAgent.js (hunk 6 above)
Comment (nanopay-arthur) [reply src/foam/u2/crunch/wizardflow/LoadCapabilitiesAgent.js]: Addressed.

File: src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+          this.tag(this.StackView.create({ data: stack, showActions: false }, x));
+        }
+      ]
+    }
+  ],
+
+  properties: [
+    {
+      class: 'foam.u2.ViewSpecWithJava',
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js]: ViewSpecWithJava has been removed. Replace with just ViewSpec.

File: src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js (hunk 1 above)
Comment (kgrgreer) [reply src/foam/nanos/crunch/AgentCapabilityJunctionDAOSummaryView.js:9]: The Approvals package could contain a refinement to AgentCapabilityJunction where it includes the DAOSummaryView as an inner-class. Then all of the code is put in the correct place but can remain on the nanopay side.

**Pull request merged**